test/perf/benchmarks/Engine.ScriptBlock.cs

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Management.Automation;
+using System.Management.Automation.Runspaces;
+using System.Runtime.InteropServices;
+using BenchmarkDotNet.Attributes;
+using MicroBenchmarks;
+
+namespace Engine
+{
+    [BenchmarkCategory(Categories.Engine, Categories.Public)]
+    public class Scripting
+    {
+        private Runspace runspace;
+        private ScriptBlock scriptBlock;
+
+        private void SetupRunspace()
+        {
+            runspace = RunspaceFactory.CreateRunspace(InitialSessionState.CreateDefault2());
+            runspace.Open();
+            Runspace.DefaultRunspace = runspace;
+        }
+
+        [ParamsSource(nameof(ValuesForScript))]
+        public string Script { get; set; }
+
+        public IEnumerable<string> ValuesForScript()
+        {
+            yield return @"'String'.GetType()";
+            yield return @"[System.IO.Path]::HasExtension('')";
+
+            // Test on COM method invocation.
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                yield return @"$sh=New-Object -ComObject Shell.Application; $sh.Namespace('c:\')";
+                yield return @"$fs=New-Object -ComObject scripting.filesystemobject; $fs.Drives";
+            }
+        }
+
+        [GlobalSetup(Target = nameof(InvokeMethod))]
+        public void GlobalSetup()
+        {
+            SetupRunspace();
+            scriptBlock = ScriptBlock.Create(Script);
+
+            // Run it once to get the C# code jitted and the script compiled.
+            // The first call to this takes relatively too long, which makes the BDN's heuristic incorrectly
+            // believe that there is no need to run many ops in each interation. However, the subsequent runs
+            // of this method is much faster than the first run, and this causes 'MinIterationTime' warnings
+            // to our benchmarks and make the benchmark results not reliable.
+            // Calling this method once in 'GlobalSetup' is a workaround. 
+            // See https://github.com/dotnet/BenchmarkDotNet/issues/837#issuecomment-828600157
+            scriptBlock.Invoke();
```

Comment:
This raises two questions:
- can we make this generic so that anybody can use this as simple pattern like `PerfMeasureSript <ScriptBlock> - Target <target>`. I guess we will measure many script blocks. This is true for any API working with scriptblocks.
- we have an internal optimization for compiling of frequently used scriptblocks. This could raise the same 'MinIterationTime' warnings.

Reply:
> simple pattern like `PerfMeasureSript <ScriptBlock> - Target <target>`

I'm not sure what pattern this is.

> Should we have a hook to disable the optimization?

The threshold for lambda compilation is 32. After applying the workaround, the number of invocations per iteration is way more than 32 (range from 200 to 200,000), and I don't see the `MinIterationTime` warning again.

We will have perf tests that intentionally avoid lambda compilation, by constructing script block that contains more than 300 statements but actually only run one statement (using `if ($false) { ... }` for instance). PowerShell never compiles script block with more than 300 statements.

Reply:
> Another hot scenario we could think is targeting "before the PR" or master git branch vs current dev branch.

This is of course doable, but I don't want to mess around with user's unstaged or uncommitted changes :)
It can be added later if it's proven to be necessary.